// inversify-express-utils/lib/base_http_controller.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
var inversify_1 = require("inversify");
var decorators_1 = require("./decorators");
var results_1 = require("./results");
var http_status_codes_1 = require("http-status-codes");
var BaseHttpController = /** @class */ (function () {
    function BaseHttpController() {
    }
    BaseHttpController.prototype.created = function (location, content) {
        return new results_1.CreatedNegotiatedContentResult(location, content, this);
    };
    BaseHttpController.prototype.conflict = function () {
        return new results_1.ConflictResult(this);
    };
    BaseHttpController.prototype.ok = function (content) {
        return content === undefined ?
            new results_1.OkResult(this) :
            new results_1.OkNegotiatedContentResult(content, this);
    };
    BaseHttpController.prototype.badRequest = function (message) {
        return message === undefined ?
            new results_1.BadRequestResult(this) :
            new results_1.BadRequestErrorMessageResult(message, this);
    };
    BaseHttpController.prototype.internalServerError = function (error) {
        return error ? new results_1.ExceptionResult(error, this) : new results_1.InternalServerErrorResult(this);
    };
    BaseHttpController.prototype.notFound = function () {
        return new results_1.NotFoundResult(this);
    };
    BaseHttpController.prototype.redirect = function (uri) {
        return new results_1.RedirectResult(uri, this);
    };
    BaseHttpController.prototype.responseMessage = function (message) {
        return new results_1.ResponseMessageResult(message, this);
    };
    BaseHttpController.prototype.statusCode = function (statusCode) {
        return new results_1.StatusCodeResult(statusCode, this);
    };
    BaseHttpController.prototype.json = function (content, statusCode) {
        if (statusCode === void 0) { statusCode = http_status_codes_1.OK; }
        return new results_1.JsonResult(content, statusCode, this);
    };
    __decorate([
        decorators_1.injectHttpContext,
        __metadata("design:type", Object)
    ], BaseHttpController.prototype, "httpContext", void 0);
    BaseHttpController = __decorate([
        inversify_1.injectable()
    ], BaseHttpController);
    return BaseHttpController;
}());
exports.BaseHttpController = BaseHttpController;
